Hide message entry when not needed

just-carlod opened the pull request:
This pr is still work in progress
- [x] Implement `status` property for Supergroup and BasicGroup
- [x]  Implement `is-blocked` and `permissions` properties for chat
- [x] Actual handling `message bar` and `restricted_label` states

User restricted in the chat:
![photo](https://user-images.githubusercontent.com/58047802/155611436-d97d0de1-eb33-47f5-93e8-26c2b755d2e2.png)
Writing not allowed in the group:
![photo](https://user-images.githubusercontent.com/58047802/156207989-cbf5da19-ccc0-4cd2-baa5-b9de4a408bc1.png)
Channel:
![photo](https://user-images.githubusercontent.com/58047802/156208042-79cb4be5-8d68-4ac8-90c5-7e6c10d14400.png)
I'll implement `Unmute` button in the future (see #258) 

fixes #35

just-carlod commented on src/session/content/chat_action_bar.rs:
```diff
@@ -287,13 +291,218 @@ impl ChatActionBar {
+            )
+            .upcast();
+
+            // Handle whether or not message bar should be shown
```
I wasn't sure if I should make it one huge expression, so I made four of them for all chat types (lines from 308 to 427)

melix99 replied:
I guess it's "fine", in the sense that gtk bindings are really complicated to use and become really big too easily. Anyway, I think that some stuff could still be simplified, I'll make a proper review soon.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
             imp.chat_action_in_cooldown.set(false);
+
+            let chat_expression = gtk::ConstantExpression::new(chat).upcast();
+            let permissions_expression = gtk::PropertyExpression::new(
```
Instead of using the classical way of creating property expression, you can "chain" them from the previous ones. In this case you can do:

```rust
let permissions_expression = chat_expression.chain_property::<Chat>("permissions").upcast();
```

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+            // Handle whether or not message bar should be shown
+            match chat.type_() {
+                ChatType::Private(data) => {
+                    let blocked_expression = gtk::PropertyExpression::new(
```
You can use the `chain_property` here too.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+                        "is-blocked",
+                    )
+                    .upcast();
+                    let user_expression = gtk::ConstantExpression::new(data).upcast();
```
You need to pass the UserType in the expression below, because an user type can be changed (e.g. Regular -> Deleted). In this case you should use the `chain_property` to retrieve it.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+                }
+                ChatType::Secret(data) => {
+                    let user = data.user();
+                    let blocked_expression = gtk::PropertyExpression::new(
```
You can use the `chain_property` here too.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+                    bindings.push(visible_expression);
+                }
+                ChatType::Supergroup(data) => {
+                    let channel_expression = gtk::PropertyExpression::new(
```
You can use the `chain_property` here too (also below).

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+                    bindings.push(visible_expression);
+                }
+                ChatType::BasicGroup(data) => {
+                    let status_expression = gtk::PropertyExpression::new(
```
You can use the `chain_property` here too.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+                            }
+                        },
+                    )
+                    .bind(&*imp.scrolled_window, "visible", Some(chat));
```
Instead of doing the same binding (and also adding it to the bindings Vec) in every single variant, you can return the `visible_expression` of each variant and then doing the actual binding after the `match`.

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
+                    )
+                    .upcast();
+                    let user_expression = gtk::ConstantExpression::new(data).upcast();
+                    let visible_expression = gtk::ClosureExpression::with_callback(
```
Instead of using the "callback" closure expression I suggest using the regular ones (`gtk::ClosureExpression::new`). You can take a look on how they work [here](https://github.com/melix99/telegrand/blob/e64043f6b597d11c4f4078759ac359f14730b27a/src/session/sidebar/row.rs#L213).

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,13 +286,218 @@ impl ChatActionBar {
         }
 
+        // Connect scrolled_window (message area) visibility with send_message_button visibility
+        self.imp()
```
I guess you wanted to do that in the `constructed()` function. But anyway I don't think this is a good idea. I think it would be better to connect the previous `visible_expression` to both the `scrolled_window` and the `send_message_button`.

marhkb commented on src/session/chat/mod.rs:
```diff
@@ -20,11 +20,15 @@ use gtk::glib;
 
+#[derive(Clone, Debug, glib::Boxed, PartialEq)]
+#[boxed_type(name = "BoxedChatPermissions")]
+pub struct BoxedChatPermissions(pub ChatPermissions);
```
```suggestion
pub(crate) struct BoxedChatPermissions(pub(crate) ChatPermissions);
```

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+                gtk::ConstantExpression::new(chat).chain_property::<Chat>("is-blocked");
+
+            // Handle whether or not message bar should be shown
+            let visible_expresison = match chat.type_() {
```
Typo. There are some more "expresison" typos in this PR.
```suggestion
            let visible_expression = match chat.type_() {
```

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+                            }
+                        }),
+                    )
+                    .upcast()
```
This and the next match arm are duplicated code. I would consider moving them in its own functions.

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+                }
+                ChatType::Supergroup(data) => {
+                    let channel_expression = gtk::ConstantExpression::new(data)
+                        .chain_property::<Supergroup>("is-channel");
```
Is an expression really necessary here? Can a chat change its channel state?
Else, I would just save it beforehand and use it in the closure.

```suggestion
                    let is_channel = data.is_channel();
```

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+                }
+            };
+
+            let message_bar_binding =
```
I think the name `message_bar_binding` binding is not very accurate. I would call it `message_bar_visibility_binding`.

just-carlod replied:
Thanks, I'm not good at keeping variable name conventions

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+
+            let message_bar_binding =
+                visible_expresison.bind(&*imp.scrolled_window, "visible", Some(chat));
+            let send_button_binding =
```
I would rather call it `send_button_visibility_binding`.

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+
+            // Handle whether or not restriction label should be shown
+            let visible_binding = gtk::ClosureExpression::new::<bool, _, _>(
+                &[chat_expression],
```
Shouldn't the "permissions_expressions" used here instead? Else, if a permission update arrives, this will never be fired.
Also the style can be simplified.

```rust
permissions_expression.chain_closure::<bool>(closure!(|/* bla */| {/* bla */});
```

marhkb commented on src/session/supergroup.rs:
```diff
@@ -123,4 +144,75 @@ impl Supergroup {
+        &self,
+        permissions_expression: gtk::PropertyExpression,
+    ) -> gtk::Expression {
+        let status_expression = Supergroup::this_expression("status");
```
```suggestion
        let status_expression = Self::this_expression("status");

```

marhkb commented on src/session/supergroup.rs:
```diff
@@ -123,4 +144,75 @@ impl Supergroup {
+                                )
+                            }
+                        }
+                    // User is restrcited, but can send messages
```
Typo

marhkb commented on src/session/supergroup.rs:
```diff
@@ -123,4 +144,75 @@ impl Supergroup {
+                                .as_days()
+                                == 0
+                            {
+                                format!(
```
This should be gettext

marhkb commented on src/session/supergroup.rs:
```diff
@@ -123,4 +144,75 @@ impl Supergroup {
+                                    date.format(&gettext("%l:%M %p")).unwrap()
+                                )
+                            } else {
+                                format!(
```
This should be gettext

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+            bindings.push(send_button_binding);
+
+            // Handle whether or not restriction label should be shown
+            let visible_binding = gtk::ClosureExpression::new::<bool, _, _>(
```
Again, I think the name is not accurate enough. Better `restriction_label_visibility_binding`.

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+                            } else if let ChatMemberStatus::Restricted(data) = data.status().0 {
+                                !data.permissions.can_send_messages
+                            } else {
+                                !chat.permissions().0.can_send_messages
```
I think you have done that check already in line 404. Could you just return true?

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+                        if let ChatType::Supergroup(data) = chat.type_() {
+                            if data.is_channel() {
+                                // TODO: implement unmute/mute button
+                                !data.is_channel()
```
Have you not checked this 2 lines before? So I think it's fine to just return false here.

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+                                // TODO: implement unmute/mute button
+                                !data.is_channel()
+                            } else if let ChatMemberStatus::Restricted(data) = data.status().0 {
+                                !data.permissions.can_send_messages
```
I need you to help me understand this. Is this value under any circumstance different from `chat.permissions().0.can_send_messages`?

just-carlod replied:
Yes they are different from each other.  `data.permissions` is set of self user permissions and `chat.permissions().0`is set of all users permissions in current chat. You're right, my code is a little junky at some parts. I'll try to refactor it so it's more efficient and more readable.

just-carlod replied:
Okay, so it's done. I hope my code looks much cleaner now

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,11 +287,183 @@ impl ChatActionBar {
+
+            bindings.push(visible_binding);
+
+            // Handle restricion_label caption
```
Typo

marhkb commented on src/session/basic_group.rs:
```diff
@@ -109,4 +127,73 @@ impl BasicGroup {
+    }
+
+    pub(crate) fn set_status(&self, status: BoxedChatMemberStatus) {
+        if self.imp().status.borrow().as_ref() == Some(&status) {
```
```suggestion
        if self.status() == status {
```

marhkb replied:
Hmm. I just saw that my change would crash, because you unwrap in `status()`. So either return an `Option` in `status` or ignore this. Depends on what @melix99 prefers.

melix99 replied:
As I wrote in the other review, this is fine for now.

marhkb commented on src/session/basic_group.rs:
```diff
@@ -109,4 +127,73 @@ impl BasicGroup {
+        self.notify("status");
+    }
+
+    pub(crate) fn restriction_expression(
```
Is this supposed to be a exact copy of Supergroup::restriction_expression`? If so, I'm not a fan of this. Also there is slight difference in the other method Supergroup:181

just-carlod replied:
> I'm not a fan of this

Sure, I can implement it in chat instead of both supergroup and basic_group

just-carlod replied:
so I did

marhkb commented on src/session/chat/mod.rs:
```diff
@@ -556,4 +602,16 @@ impl Chat {
+    }
+
+    pub(crate) fn set_permissions(&self, permissions: BoxedChatPermissions) {
+        if self.imp().permissions.borrow().as_ref() == Some(&permissions) {
```
```suggestion
        if self.permissions() == permissions {
```

marhkb replied:
Hmm. I just saw that my change would crash, because you unwrap in `permissions()`. So either return an `Option` in `permissions` or ignore this. Depends on what @melix99 prefers.

melix99 replied:
This is a bit tricky in the current state because all the properties are None by default. This will change with the new tdlib-gobject library, but for now I guess this is fine, this is also what we currently do in some other cases.

marhkb commented on src/session/basic_group.rs:
```diff
@@ -109,4 +127,73 @@ impl BasicGroup {
     }
+
+    pub(crate) fn status(&self) -> BoxedChatMemberStatus {
+        self.imp().status.borrow().as_ref().unwrap().to_owned()
```
```suggestion
        self.imp().status.borrow().to_owned().unwrap()
```

marhkb commented on src/session/chat/mod.rs:
```diff
@@ -556,4 +602,16 @@ impl Chat {
     }
+
+    pub(crate) fn permissions(&self) -> BoxedChatPermissions {
+        self.imp().permissions.borrow().as_ref().unwrap().to_owned()
```
```suggestion
        self.imp().permissions.borrow().to_owned().unwrap()
```

marhkb replied:
Forgot this one?

EDIT: 
github doesn't include the previous thread here, as it seems
It can be shortened
```rust
self.imp().permissions.borrow().to_owned().unwrap()
```

melix99 commented on src/session/chat/mod.rs:
```diff
@@ -544,4 +594,102 @@ impl Chat {
+        self.notify("permissions");
+    }
+
+    pub(crate) fn restriction_expression(&self) -> gtk::Expression {
```
This needs to be moved in the new expressions module.

melix99 commented on src/session/chat/mod.rs:
```diff
@@ -544,4 +594,102 @@ impl Chat {
+    }
+}
+
+fn restriction_expression_label(
```
This needs to be moved in the new expressions module.

marhkb commented on src/expressions.rs:
```diff
@@ -30,3 +36,90 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
+                    restriction_label_expression(status, chat_permissions)
+                },
+            )
+            .upcast()
```
This code and the code in the next match arm is duplicated. It can be put in one function if you generalize a bit.

marhkb commented on src/expressions.rs:
```diff
@@ -30,3 +36,90 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
+    if chat_permissions.can_send_messages {
+        if let ChatMemberStatus::Restricted(status) = status {
+            if !status.permissions.can_send_messages {
+                if !status.restricted_until_date == 0 {
```
```suggestion
                if status.restricted_until_date != 0 {
```

marhkb commented on src/expressions.rs:
```diff
@@ -30,3 +36,90 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
+                        .difference(&date)
+                        .as_days()
+                        == 0
+                    {
```
I would prefer this `if` condition directly in the `gettext` macro. This would save a bit of code.

marhkb commented on src/session/basic_group.rs:
```diff
@@ -45,6 +48,15 @@ mod imp {
+                    glib::ParamSpecBoxed::new(
+                        "status",
+                        "Status",
+                        "Self user status in this basic group",
```
```suggestion
                        "Own user status in this basic group",
```

marhkb commented on src/session/chat/mod.rs:
```diff
@@ -17,12 +17,20 @@ pub(crate) use self::sponsored_message::SponsoredMessage;
+#[boxed_type(name = "BoxedChatMemberStatus")]
+pub(crate) struct BoxedChatMemberStatus(pub(crate) ChatMemberStatus);
+
+#[derive(Clone, Debug, glib::Boxed, PartialEq)]
```
```suggestion
#[derive(Clone, Debug, PartialEq, glib::Boxed)]
```

marhkb commented on src/session/chat/mod.rs:
```diff
@@ -126,6 +136,15 @@ mod imp {
+                    glib::ParamSpecBoolean::new(
+                        "is-blocked",
+                        "Is blocked",
+                        "Whether or not this chat is blocked by the user",
```
"for" instead of "by"?
```suggestion
                        "Whether this chat is blocked for the user",
```

marhkb commented on src/session/chat/mod.rs:
```diff
@@ -226,6 +245,15 @@ mod imp {
+                        BoxedChatPermissions::static_type(),
+                        glib::ParamFlags::READWRITE
+                            | glib::ParamFlags::EXPLICIT_NOTIFY
+                            | glib::ParamFlags::CONSTRUCT,
```
```suggestion
                            | glib::ParamFlags::CONSTRUCT
                            | glib::ParamFlags::EXPLICIT_NOTIFY,

```

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,14 +286,167 @@ impl ChatActionBar {
+            let chat_expression = gtk::ConstantExpression::new(chat).upcast();
+            let permissions_expression = chat_expression.chain_property::<Chat>("permissions");
+            let blocked_expression =
+                gtk::ConstantExpression::new(chat).chain_property::<Chat>("is-blocked");
```
```suggestion
            let permissions_expression = Chat::this_expression("permissions");
            let blocked_expression = Chat::this_expression("is-blocked");
```

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,14 +286,167 @@ impl ChatActionBar {
+                gtk::ConstantExpression::new(chat).chain_property::<Chat>("is-blocked");
+
+            // Handle whether or not message bar should be shown
+            let message_bar_visibility_expresison = match chat.type_() {
```
```suggestion
            let message_bar_visibility_expression = match chat.type_() {
```

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,14 +286,167 @@ impl ChatActionBar {
+                        gtk::ConstantExpression::new(data.user()).chain_property::<User>("type");
+                    message_bar_visibility_in_private_chats(blocked_expression, type_expression)
+                }
+                ChatType::Supergroup(data) if data.is_channel() => {
```
The match guard should be removed here and the match arm should be merged with the next match arm. Then, the `if` condition can live inside that match arm instead of being a guard.

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -282,14 +286,167 @@ impl ChatActionBar {
 
             imp.chat_action_in_cooldown.set(false);
+
+            let chat_expression = gtk::ConstantExpression::new(chat).upcast();
```
I think you should set up the expressions within the `constructed` phase and use `let chat_expression = Self::this_expression("chat")` as an anchor. In this way, you don't need to save the bindings and unwatch them.

See
https://github.com/melix99/telegrand/blob/1b438b5d1657302a820934f3e2ab82a36dd905e9/src/session/content/chat_history.rs#L133

just-carlod replied:
Is this still required? since this variable seems to be unused, everything works without it (or at least on first glance)
```rust
let permissions_expression = Chat::this_expression("permissions");
let blocked_expression = Chat::this_expression("is-blocked");
```

marhkb commented on src/session/content/chat_action_bar.rs:
```diff
@@ -283,14 +288,164 @@ impl ChatActionBar {
+                    false
+                }))
+                .upcast()
+                .bind(&*imp.restriction_label, "visible", Some(chat));
```
I really don't like these deep nested returns. I consider them bad style in an expression based language like rust and they make code harder to read.
It could be solved like this.

 ```suggestion
            let restriction_label_visibility_binding = permissions_expression
                .chain_closure::<bool>(closure!(|chat: Chat, permissions: BoxedChatPermissions| {
                    if permissions.0.can_send_messages {
                        match chat.type_() {
                            ChatType::Supergroup(data) => Some(data.status().0),
                            ChatType::BasicGroup(data) => Some(data.status().0),
                            _ => None,
                        }
                        .map(|status| match status {
                            ChatMemberStatus::Restricted(status) => {
                                !status.permissions.can_send_messages
                            }
                            _ => false,
                        })
                        .unwrap_or(false)
                    } else {
                        match chat.type_() {
                            ChatType::Supergroup(data) if !data.is_channel() => {
                                !matches!(
                                    data.status().0,
                                    ChatMemberStatus::Creator(_)
                                        | ChatMemberStatus::Administrator(_)
                                )
                            }
                            ChatType::BasicGroup(data) => {
                                !matches!(
                                    data.status().0,
                                    ChatMemberStatus::Creator(_)
                                        | ChatMemberStatus::Administrator(_)
                                )
                            }
                            _ => false,
                        }
                    }
                }))
                .upcast()
                .bind(&*imp.restriction_label, "visible", Some(chat));
```

marhkb commented on src/expressions.rs:
```diff
@@ -30,3 +35,84 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
+
+pub(crate) fn restriction_expression(chat: &Chat) -> gtk::Expression {
+    let chat_expression = gtk::ConstantExpression::new(chat).upcast();
+    let permissions_expression = chat_expression.chain_property::<Chat>("permissions");
```
```suggestion
    let permissions_expression = Chat::this_expression("permissions");
```

marhkb commented on src/expressions.rs:
```diff
@@ -30,3 +35,84 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
+                Some(gtk::ConstantExpression::new(data)),
+                "status",
+            );
+            restriction_label_expression(&member_status_expression, &permissions_expression)
```
This better, but can still be improved if you work with type parameters

Here is my suggestion (I also improved the `deep return` spaghetti code)

```rust
pub(crate) fn restriction_expression(chat: &Chat) -> gtk::Expression {
    // chat type is (usually) constant so we don't need to update it
    match chat.type_() {
        ChatType::Supergroup(data) if !data.is_channel() => {
            restriction_label_expression::<Supergroup, _>(data)
        }
        ChatType::BasicGroup(data) => restriction_label_expression::<BasicGroup, _>(data),
        _ => gtk::ConstantExpression::new(&"").upcast(),
    }
}

fn restriction_label_expression<T: glib::StaticType, V: glib::ToValue>(
    value: &V,
) -> gtk::Expression {
    let member_status_expression = gtk::PropertyExpression::new(
        T::static_type(),
        Some(gtk::ConstantExpression::new(value)),
        "status",
    );
    let permissions_expression = Chat::this_expression("permissions");

    gtk::ClosureExpression::new::<String, _, _>(
        &[member_status_expression, permissions_expression],
        closure!(|_: glib::Object, status: BoxedChatMemberStatus, chat_permissions: BoxedChatPermissions| {
            if chat_permissions.0.can_send_messages {
                return String::new()
            }

            match status.0 {
                ChatMemberStatus::Restricted(status) if !status.permissions.can_send_messages => {
                    if status.restricted_until_date == 0 {
                        gettext("The admins of this group have restricted you from writing here")
                    } else {
                        let date =
                            glib::DateTime::from_unix_utc(status.restricted_until_date.into()).unwrap();

                        gettext!(
                            "The admins of this group have restricted you from writing here until {}",
                            date.format(&if glib::DateTime::now_local()
                                .unwrap()
                                .difference(&date)
                                .as_days()
                                == 0
                            {
                                // Translators: This is a hours: minutes time format used in restriction label
                                gettext("%l:%M %p")
                            } else {
                                // Translators: This is a full date format used in restriction label
                                gettext("%B %e, %Y %l:%M %p")
                            })
                            .unwrap()
                            .to_string()
                        )
                    }
                }
                ChatMemberStatus::Creator(_) | ChatMemberStatus::Administrator(_) => {
                    gettext("Writing messages isn't allowed in this group.")
                }
                _ => String::new(),
            }
        }),
    )
    .upcast()
}

```

just-carlod replied:
Thanks for your help, there was an issue in your code:
`` ChatMemberStatus::Creator(_) | ChatMemberStatus::Administrator(_) => {
                    gettext("Writing messages isn't allowed in this group.")
                }``

- admins always have permissions to write messages

but I corrected that part

just-carlod replied:
I believe it should be fine now

marhkb commented on src/expressions.rs:
```diff
@@ -30,3 +36,68 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
+                        } else {
+                            String::new()
+                        }
+
```
The formatting is wrong here, as rustfmt sometimes fails in macros. In such cases, I copy and paste the code to a location somewhere outside the macro with the same level of indentation, run rustfmt, and copy it back.

just-carlod replied:
Thanks for your tip, I'll fix it later

just-carlod committed:
Implement `status` property for `Supergroup` and `BasicGroup`

just-carlod committed:
chat: Implement `is-blocked` and `permissions` properties

just-carlod committed:
chat_action_bar: Hide message entry when not needed

Hide message entry when user is not allowed to send messages in current chat.
Always hide message bar in a private chat with deleted account.

Co-authored-by: Marcus Behrendt <marcus.behrendt.86@gmail.com>

just-carlod committed:
Introduce restriction label

Introduce restriction label which displays the date to which user is restricted.
It hides automatically when user is not restricted

Co-authored-by: Marcus Behrendt <marcus.behrendt.86@gmail.com>

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -2,18 +2,21 @@ use anyhow::anyhow;
+use glib::{clone, closure};
 use gtk::prelude::*;
 use gtk::subclass::prelude::*;
+
```
```suggestion
```

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -384,3 +502,44 @@ async fn save_stream_to_file(
+
+fn message_bar_visibility_in_private_chats(
+    blocked_expression: gtk::PropertyExpression,
+    type_expression: gtk::PropertyExpression,
```
```suggestion
    is_blocked_expression: gtk::PropertyExpression,
    user_type_expression: gtk::PropertyExpression,
```

melix99 commented on src/session/content/chat_action_bar.rs:
```diff
@@ -384,3 +502,44 @@ async fn save_stream_to_file(
+
+fn message_bar_visibility_in_groups(
+    permissions_expression: gtk::PropertyExpression,
+    status_expression: gtk::PropertyExpression,
```
```suggestion
    member_status_expression: gtk::PropertyExpression,
```

melix99 commented on src/expressions.rs:
```diff
@@ -30,3 +36,67 @@ pub(crate) fn user_full_name(user_expression: &gtk::Expression) -> gtk::Expressi
 }
+
+pub(crate) fn restriction_expression(chat: &Chat) -> gtk::Expression {
+    // chat type is (usually) constant so we don't need to update it
```
```suggestion
```

I think it's obvious enough.

melix99 commented on data/resources/ui/content-chat-action-bar.ui:
```diff
@@ -33,5 +33,14 @@
       </object>
     </child>
+    <child>
+      <object class="GtkLabel" id="restriction_label">
```
I suggest adding `<property name="justify">center</property>`, this way longer labels look a bit better.

just-carlod replied:
I also found `<property name="visible">False</property>` to be useless since we have binding for that property
